src/python_bluesky_taskgraph/core/task.py

```diff
@@ -147,3 +150,29 @@ def _overwrite_results(self, results: List[Any] = None) -> None:
 
     def get_results(self, keys: List[str]) -> Dict[str, Any]:
         return {k: v for (k, v) in zip(keys, self._results) if k is not None}
+
+
+def run_stage_decorator(
+    func: Callable[[InputType], TaskOutput]
+) -> Callable[[InputType], TaskOutput]:
+    def decorated_func(args: InputType):
+        devices = {device for device in astuple(args) if isinstance(device, Device)}
+        for device in devices:
+            yield from stage(device)
+        yield from func(args)
+        for device in devices:
+            yield from unstage(device)
+
+    return decorated_func
+
+
+def task_stage_decorator(
+    func: Callable[..., BlueskyTask[InputType]]
+) -> Callable[..., BlueskyTask[InputType]]:
+    def wrapper_stage_decorator(*args, **kwargs) -> BlueskyTask[InputType]:
+        task: BlueskyTask = func(*args, **kwargs)
+        # Prevents MyPy complaints about assigning to a function
+        task.__setattr__("_run_task", run_stage_decorator(task._run_task))
```

Comment:
Not too bothered about this, but alternative is to make a wrapper task

Reply:
Harder to test, as the passing of the status between a wrapping mock and the underlying task leads to problems passing completion around. I do think it's a neater way of doing it though, implementation similar to ConditionalTask.